[debug dump util]FDB debug dump util changes

dgsudharsan opened the pull request:
Signed-off-by: Sudharsan Dhamal Gopalarathnam <sudharsand@nvidia.com>

<!--
    Please make sure you've read and understood our contributing guidelines:
    https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

    ** Make sure all your commits include a signature generated with `git commit -s` **

    If this is a bug fix, make sure your description includes "closes #xxxx",
    "fixes #xxxx" or "resolves #xxxx" so that GitHub automatically closes the related
    issue when the PR is merged.

    If you are adding/modifying/removing any command or utility script, please also
    make sure to add/modify/remove any unit tests from the tests
    directory as appropriate.

    If you are modifying or removing an existing 'show', 'config' or 'sonic-clear'
    subcommand, or you are adding a new subcommand, please make sure you also
    update the Command Line Reference Guide (doc/Command-Reference.md) to reflect
    your changes.

    Please provide the following information:
-->

#### What I did
Added FDB debug dump module

#### How I did it
Used debug dump infra to add the FDB debug dump utility

#### How to verify it
Added Unit test cases to verify it.

#### Previous command output (if the output of a command-line utility has changed)

#### New command output (if the output of a command-line utility has changed)

```
root@sonic:~# dump state fdb all
{
    "Vlan50:04:3f:72:e3:70:08": {
        "APPL_DB": {
            "keys": [],
            "tables_not_found": []
        },
        "ASIC_DB": {
            "keys": [
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_BRIDGE_PORT:oid:0x3a000000000d23": {
                        "SAI_BRIDGE_PORT_ATTR_ADMIN_STATE": "true",
                        "SAI_BRIDGE_PORT_ATTR_FDB_LEARNING_MODE": "SAI_BRIDGE_PORT_FDB_LEARNING_MODE_HW",
                        "SAI_BRIDGE_PORT_ATTR_PORT_ID": "oid:0x100000000093e",
                        "SAI_BRIDGE_PORT_ATTR_TYPE": "SAI_BRIDGE_PORT_TYPE_PORT"
                    }
                },
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_FDB_ENTRY:{\"bvid\":\"oid:0x26000000000d22\",\"mac\":\"04:3F:72:E3:70:08\",\"switch_id\":\"oid:0x21000000000000\"}": {
                        "SAI_FDB_ENTRY_ATTR_BRIDGE_PORT_ID": "oid:0x3a000000000d23",
                        "SAI_FDB_ENTRY_ATTR_PACKET_ACTION": "SAI_PACKET_ACTION_FORWARD",
                        "SAI_FDB_ENTRY_ATTR_TYPE": "SAI_FDB_ENTRY_TYPE_DYNAMIC"
                    }
                }
            ],
            "tables_not_found": [],
            "vidtorid": {
                "oid:0x3a000000000d23": "oid:0x20000003a"
            }
        },
        "STATE_DB": {
            "keys": [
                {
                    "FDB_TABLE|Vlan50:04:3f:72:e3:70:08": {
                        "port": "Ethernet0",
                        "type": "dynamic"
                    }
                }
            ],
            "tables_not_found": []
        }
    },
    "Vlan10:04:3f:72:ce:80:8b": {
        "APPL_DB": {
            "keys": [],
            "tables_not_found": []
        },
        "ASIC_DB": {
            "keys": [
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_BRIDGE_PORT:oid:0x3a000000000d18": {
                        "SAI_BRIDGE_PORT_ATTR_ADMIN_STATE": "true",
                        "SAI_BRIDGE_PORT_ATTR_FDB_LEARNING_MODE": "SAI_BRIDGE_PORT_FDB_LEARNING_MODE_HW",
                        "SAI_BRIDGE_PORT_ATTR_PORT_ID": "oid:0x100000000053f",
                        "SAI_BRIDGE_PORT_ATTR_TYPE": "SAI_BRIDGE_PORT_TYPE_PORT"
                    }
                },
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_FDB_ENTRY:{\"bvid\":\"oid:0x26000000000d20\",\"mac\":\"04:3F:72:CE:80:8B\",\"switch_id\":\"oid:0x21000000000000\"}": {
                        "SAI_FDB_ENTRY_ATTR_BRIDGE_PORT_ID": "oid:0x3a000000000d18",
                        "SAI_FDB_ENTRY_ATTR_PACKET_ACTION": "SAI_PACKET_ACTION_FORWARD",
                        "SAI_FDB_ENTRY_ATTR_TYPE": "SAI_FDB_ENTRY_TYPE_DYNAMIC"
                    }
                }
            ],
            "tables_not_found": [],
            "vidtorid": {
                "oid:0x3a000000000d18": "oid:0x10000003a"
            }
        },
        "STATE_DB": {
            "keys": [
                {
                    "FDB_TABLE|Vlan10:04:3f:72:ce:80:8b": {
                        "port": "Ethernet252",
                        "type": "dynamic"
                    }
                }
            ],
            "tables_not_found": []
        }
    },
    "Vlan690:04:3f:72:ce:80:8b": {
        "APPL_DB": {
            "keys": [],
            "tables_not_found": []
        },
        "ASIC_DB": {
            "keys": [
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_BRIDGE_PORT:oid:0x3a000000000d18": {
                        "SAI_BRIDGE_PORT_ATTR_ADMIN_STATE": "true",
                        "SAI_BRIDGE_PORT_ATTR_FDB_LEARNING_MODE": "SAI_BRIDGE_PORT_FDB_LEARNING_MODE_HW",
                        "SAI_BRIDGE_PORT_ATTR_PORT_ID": "oid:0x100000000053f",
                        "SAI_BRIDGE_PORT_ATTR_TYPE": "SAI_BRIDGE_PORT_TYPE_PORT"
                    }
                },
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_FDB_ENTRY:{\"bvid\":\"oid:0x26000000000d1c\",\"mac\":\"04:3F:72:CE:80:8B\",\"switch_id\":\"oid:0x21000000000000\"}": {
                        "SAI_FDB_ENTRY_ATTR_BRIDGE_PORT_ID": "oid:0x3a000000000d18",
                        "SAI_FDB_ENTRY_ATTR_PACKET_ACTION": "SAI_PACKET_ACTION_FORWARD",
                        "SAI_FDB_ENTRY_ATTR_TYPE": "SAI_FDB_ENTRY_TYPE_DYNAMIC"
                    }
                }
            ],
            "tables_not_found": [],
            "vidtorid": {
                "oid:0x3a000000000d18": "oid:0x10000003a"
            }
        },
        "STATE_DB": {
            "keys": [
                {
                    "FDB_TABLE|Vlan690:04:3f:72:ce:80:8b": {
                        "port": "Ethernet252",
                        "type": "dynamic"
                    }
                }
            ],
            "tables_not_found": []
        }
    },
    "Vlan40:04:3f:72:e3:70:09": {
        "APPL_DB": {
            "keys": [],
            "tables_not_found": []
        },
        "ASIC_DB": {
            "keys": [
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_BRIDGE_PORT:oid:0x3a000000000d16": {
                        "SAI_BRIDGE_PORT_ATTR_ADMIN_STATE": "true",
                        "SAI_BRIDGE_PORT_ATTR_FDB_LEARNING_MODE": "SAI_BRIDGE_PORT_FDB_LEARNING_MODE_HW",
                        "SAI_BRIDGE_PORT_ATTR_PORT_ID": "oid:0x2000000000d13",
                        "SAI_BRIDGE_PORT_ATTR_TYPE": "SAI_BRIDGE_PORT_TYPE_PORT"
                    }
                },
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_FDB_ENTRY:{\"bvid\":\"oid:0x26000000000d15\",\"mac\":\"04:3F:72:E3:70:09\",\"switch_id\":\"oid:0x21000000000000\"}": {
                        "SAI_FDB_ENTRY_ATTR_BRIDGE_PORT_ID": "oid:0x3a000000000d16",
                        "SAI_FDB_ENTRY_ATTR_PACKET_ACTION": "SAI_PACKET_ACTION_FORWARD",
                        "SAI_FDB_ENTRY_ATTR_TYPE": "SAI_FDB_ENTRY_TYPE_DYNAMIC"
                    }
                }
            ],
            "tables_not_found": [],
            "vidtorid": {
                "oid:0x3a000000000d16": "oid:0x3a"
            }
        },
        "STATE_DB": {
            "keys": [
                {
                    "FDB_TABLE|Vlan40:04:3f:72:e3:70:09": {
                        "port": "PortChannel0002",
                        "type": "dynamic"
                    }
                }
            ],
            "tables_not_found": []
        }
    },
    "Vlan691:04:3f:72:ce:80:8b": {
        "APPL_DB": {
            "keys": [],
            "tables_not_found": []
        },
        "ASIC_DB": {
            "keys": [
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_BRIDGE_PORT:oid:0x3a000000000d18": {
                        "SAI_BRIDGE_PORT_ATTR_ADMIN_STATE": "true",
                        "SAI_BRIDGE_PORT_ATTR_FDB_LEARNING_MODE": "SAI_BRIDGE_PORT_FDB_LEARNING_MODE_HW",
                        "SAI_BRIDGE_PORT_ATTR_PORT_ID": "oid:0x100000000053f",
                        "SAI_BRIDGE_PORT_ATTR_TYPE": "SAI_BRIDGE_PORT_TYPE_PORT"
                    }
                },
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_FDB_ENTRY:{\"bvid\":\"oid:0x26000000000d1e\",\"mac\":\"04:3F:72:CE:80:8B\",\"switch_id\":\"oid:0x21000000000000\"}": {
                        "SAI_FDB_ENTRY_ATTR_BRIDGE_PORT_ID": "oid:0x3a000000000d18",
                        "SAI_FDB_ENTRY_ATTR_PACKET_ACTION": "SAI_PACKET_ACTION_FORWARD",
                        "SAI_FDB_ENTRY_ATTR_TYPE": "SAI_FDB_ENTRY_TYPE_DYNAMIC"
                    }
                }
            ],
            "tables_not_found": [],
            "vidtorid": {
                "oid:0x3a000000000d18": "oid:0x10000003a"
            }
        },
        "STATE_DB": {
            "keys": [
                {
                    "FDB_TABLE|Vlan691:04:3f:72:ce:80:8b": {
                        "port": "Ethernet252",
                        "type": "dynamic"
                    }
                }
            ],
            "tables_not_found": []
        }
    },
    "Vlan69:04:3f:72:e3:70:09": {
        "APPL_DB": {
            "keys": [],
            "tables_not_found": []
        },
        "ASIC_DB": {
            "keys": [
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_BRIDGE_PORT:oid:0x3a000000000d16": {
                        "SAI_BRIDGE_PORT_ATTR_ADMIN_STATE": "true",
                        "SAI_BRIDGE_PORT_ATTR_FDB_LEARNING_MODE": "SAI_BRIDGE_PORT_FDB_LEARNING_MODE_HW",
                        "SAI_BRIDGE_PORT_ATTR_PORT_ID": "oid:0x2000000000d13",
                        "SAI_BRIDGE_PORT_ATTR_TYPE": "SAI_BRIDGE_PORT_TYPE_PORT"
                    }
                },
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_FDB_ENTRY:{\"bvid\":\"oid:0x26000000000d1a\",\"mac\":\"04:3F:72:E3:70:09\",\"switch_id\":\"oid:0x21000000000000\"}": {
                        "SAI_FDB_ENTRY_ATTR_BRIDGE_PORT_ID": "oid:0x3a000000000d16",
                        "SAI_FDB_ENTRY_ATTR_PACKET_ACTION": "SAI_PACKET_ACTION_FORWARD",
                        "SAI_FDB_ENTRY_ATTR_TYPE": "SAI_FDB_ENTRY_TYPE_DYNAMIC"
                    }
                }
            ],
            "tables_not_found": [],
            "vidtorid": {
                "oid:0x3a000000000d16": "oid:0x3a"
            }
        },
        "STATE_DB": {
            "keys": [
                {
                    "FDB_TABLE|Vlan69:04:3f:72:e3:70:09": {
                        "port": "PortChannel0002",
                        "type": "dynamic"
                    }
                }
            ],
            "tables_not_found": []
        }
    },
    "Vlan101:04:3f:72:e3:70:09": {
        "APPL_DB": {
            "keys": [],
            "tables_not_found": []
        },
        "ASIC_DB": {
            "keys": [
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_BRIDGE_PORT:oid:0x3a000000000d16": {
                        "SAI_BRIDGE_PORT_ATTR_ADMIN_STATE": "true",
                        "SAI_BRIDGE_PORT_ATTR_FDB_LEARNING_MODE": "SAI_BRIDGE_PORT_FDB_LEARNING_MODE_HW",
                        "SAI_BRIDGE_PORT_ATTR_PORT_ID": "oid:0x2000000000d13",
                        "SAI_BRIDGE_PORT_ATTR_TYPE": "SAI_BRIDGE_PORT_TYPE_PORT"
                    }
                },
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_FDB_ENTRY:{\"bvid\":\"oid:0x26000000000d28\",\"mac\":\"04:3F:72:E3:70:09\",\"switch_id\":\"oid:0x21000000000000\"}": {
                        "SAI_FDB_ENTRY_ATTR_BRIDGE_PORT_ID": "oid:0x3a000000000d16",
                        "SAI_FDB_ENTRY_ATTR_PACKET_ACTION": "SAI_PACKET_ACTION_FORWARD",
                        "SAI_FDB_ENTRY_ATTR_TYPE": "SAI_FDB_ENTRY_TYPE_DYNAMIC"
                    }
                }
            ],
            "tables_not_found": [],
            "vidtorid": {
                "oid:0x3a000000000d16": "oid:0x3a"
            }
        },
        "STATE_DB": {
            "keys": [
                {
                    "FDB_TABLE|Vlan101:04:3f:72:e3:70:09": {
                        "port": "PortChannel0002",
                        "type": "dynamic"
                    }
                }
            ],
            "tables_not_found": []
        }
    },
    "Vlan40:04:3f:72:ce:80:8b": {
        "APPL_DB": {
            "keys": [],
            "tables_not_found": []
        },
        "ASIC_DB": {
            "keys": [
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_BRIDGE_PORT:oid:0x3a000000000d18": {
                        "SAI_BRIDGE_PORT_ATTR_ADMIN_STATE": "true",
                        "SAI_BRIDGE_PORT_ATTR_FDB_LEARNING_MODE": "SAI_BRIDGE_PORT_FDB_LEARNING_MODE_HW",
                        "SAI_BRIDGE_PORT_ATTR_PORT_ID": "oid:0x100000000053f",
                        "SAI_BRIDGE_PORT_ATTR_TYPE": "SAI_BRIDGE_PORT_TYPE_PORT"
                    }
                },
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_FDB_ENTRY:{\"bvid\":\"oid:0x26000000000d15\",\"mac\":\"04:3F:72:CE:80:8B\",\"switch_id\":\"oid:0x21000000000000\"}": {
                        "SAI_FDB_ENTRY_ATTR_BRIDGE_PORT_ID": "oid:0x3a000000000d18",
                        "SAI_FDB_ENTRY_ATTR_PACKET_ACTION": "SAI_PACKET_ACTION_FORWARD",
                        "SAI_FDB_ENTRY_ATTR_TYPE": "SAI_FDB_ENTRY_TYPE_DYNAMIC"
                    }
                }
            ],
            "tables_not_found": [],
            "vidtorid": {
                "oid:0x3a000000000d18": "oid:0x10000003a"
            }
        },
        "STATE_DB": {
            "keys": [
                {
                    "FDB_TABLE|Vlan40:04:3f:72:ce:80:8b": {
                        "port": "Ethernet252",
                        "type": "dynamic"
                    }
                }
            ],
            "tables_not_found": []
        }
    }
}
root@sonic:~#
```

dgsudharsan commented on dump/plugins/fdb.py:
```diff
@@ -0,0 +1,103 @@
+        ret = self.match_engine.fetch(req)
+        self.add_to_ret_template(req.table, req.db, ret["keys"], ret["error"])
+
+    def init_fdb_appl_info(self, fdb_name):
```
The idea to not add APP_DB tables when missing is on purpose. In most scenarios FDB will be present only in ASIC and STATE_DB. Only in EVPN and MCLAG they are supposed to be programmed through control plane. If we display table not found for every hardware learnt entry it might create a wrong impression as something is missing particularly in fdb scenario.

vivekreddynv replied:
In that case, instead of using len(ret['keys']) != 0,  the add_to_tables_not_found argument could be set to False could set the `add_to_ret_template(self, table, db, keys, err, add_to_tables_not_found=True)`

keboliu commented:
there is a configuration for the FDB aging time, should it also be covered?

```
    {
        "SWITCH_TABLE:switch": {
            "fdb_aging_time":"120",
            "ecmp_hash_seed": "{{ hash_seed }}",
            "lag_hash_seed": "{{ hash_seed }}"
        },
        "OP": "SET"
    }
```

keboliu commented on tests/dump_input/fdb/appl_db.json:
```diff
@@ -0,0 +1,15 @@
+{
```
suggest to cover the "static" mac entry case

dgsudharsan replied:
Changed the test case into static entry since loading via appl_db pretty much is for static mac entry scenario.

dgsudharsan commented:
> 1. there is a configuration for the FDB aging time, should it also be covered?
> 
> ```
>     {
>         "SWITCH_TABLE:switch": {
>             "fdb_aging_time":"120",
>             "ecmp_hash_seed": "{{ hash_seed }}",
>             "lag_hash_seed": "{{ hash_seed }}"
>         },
>         "OP": "SET"
>     }
> ```
> 
[Sudharsan] This debug dump is for fdb entry. When we add debug dump for switch we can include to cover all global level settings.
> 2. Have you considered the Multi-ASIC scenario?
[Sudharsan] The namespace is incorporated into debug dump framework using which you can provide the namespace of the DB corresponding to ASIC that needs to be dumped (  -n, --namespace TEXT  Dump the redis-state for this namespace.  [default: DEFAULT_NAMESPACE])
> 3. Have the static mac configuration also be covered in the current implementation?
For static entry I believe it is populated using swssconfig and added into APPL_DB under FDB_TABLE. This is handled in the implementation. I have also updated test case so it reflects.

dgsudharsan commented:
@keboliu  I have pushed some changes in test and also answered your questions. Let me know if it is good.

vivekreddynv commented on dump/plugins/fdb.py:
```diff
@@ -0,0 +1,103 @@
+        self.init_state_fdb_info(fdb_entry)
+        return self.ret_temp
+
+    def add_to_ret_template(self, table, db, keys, err):
```
This can be removed. the method has been added to the parent class. https://github.com/Azure/sonic-utilities/commit/4e132c1c86c515076815a64cf22aaf8e0101de84

vivekreddynv commented on tests/dump_tests/module_tests/fdb_test.py:
```diff
@@ -0,0 +1,196 @@
+dedicated_dbs['STATE_DB'] = os.path.join(fdb_files_path, "state_db.json")
+
+
+def populate_mock(db, db_names):
```
This can be removed and can be imported from dump.helper. Use this ref: https://github.com/Azure/sonic-utilities/commit/4e132c1c86c515076815a64cf22aaf8e0101de84 to modify the test based on the refactored design

vivekreddynv commented on dump/plugins/fdb.py:
```diff
@@ -0,0 +1,90 @@
+    def init_fdb_appl_info(self, fdb_name):
+        req = MatchRequest(db="APPL_DB", table="FDB_TABLE", key_pattern=fdb_name, ns=self.ns)
+        ret = self.match_engine.fetch(req)
+        self.add_to_ret_template(req.table, req.db, ret["keys"], ret["error"], false)
```
This should be `False`
